File: 프로그래머스/2/468377. 힌트 스테이지/힌트 스테이지.py
def solution(cost, hint):
    answer = float('inf')
    
    n = len(cost) # 스테이지 개수
    hint_count = [0] * n
    
    def dfs(idx, total_bundle_cost):
        nonlocal answer
        
        if idx == n - 1: # 모든 힌트 번들 구매 여부 결정끝(dfs 종료 조건)
            total = total_bundle_cost
            
            for stage in range(n):
                cnt = min(hint_count[stage], len(cost[stage]) - 1)
                total += cost[stage][cnt]
            
            answer = min(answer, total)
            return
        
        # idx번 힌트 번들 구매 X
        dfs(idx + 1, total_bundle_cost)
        
        # idx번 힌트 번들 구매 O
        bundle = hint[idx]
        price = bundle[0] # 힌트 번들 구매 가격
        stages = bundle[1:] # 힌트권 번호들
        
        for stage in stages:
            hint_count[stage - 1] += 1
            
        dfs(idx+1, total_bundle_cost + price)
        
        # 원상복구
        for stage in stages:
            hint_count[stage - 1] -= 1
    
    dfs(0, 0)
    return answer

File: 프로그래머스/2/468377. 힌트 스테이지/힌트 스테이지.java
class Solution {
    int answer;
    int n;
    int[] hintCount;
    
    public int solution(int[][] cost, int[][] hint) {
        answer = Integer.MAX_VALUE;
        n = cost.length;
        hintCount = new int[n];
        
        dfs(0, 0, cost, hint);
        
        return answer;
    }
    
    private void dfs(int idx, int bundleCost, int[][] cost, int[][] hint) {
        if (idx == n - 1) {
            int totalCost = bundleCost;
            
            for (int stage = 0; stage < n; stage++) {
                int cnt = hintCount[stage];
                cnt = Math.min(cnt, cost[stage].length - 1);
                totalCost += cost[stage][cnt];
            }
            
            answer = Math.min(answer, totalCost);
            return;
        }
        
        // 번들을 사지 않음
        dfs(idx + 1, bundleCost, cost, hint);
        
        // idx번째 번들을 삼
        int price = hint[idx][0];
        for (int i = 1; i < hint[idx].length; i++) {
            int stage = hint[idx][i] - 1;
            hintCount[stage]++; // 해당 스테이지 힌트권 사용 수 up
        }
        dfs(idx + 1, bundleCost + price, cost, hint);
        
        // 백트래킹
        for (int i = 1; i < hint[idx].length; i++) {
            int stage = hint[idx][i] - 1;
            hintCount[stage]--;
        }
    }
}